### PR TITLE
Perf: URL refactor

### PR DESCRIPTION
**Pull request by ShrutiVerma97**

### CODE REVIEW

File: example/src/main/scala/example/PlainTextBenchmarkServer.scala [hunk 1]
```diff
@@ -28,7 +28,9 @@ object Main extends App {
       .exitCode
   }
 
-  private def app(response: Response) = Http.response(response)
+  private def app(response: Response) = Http.collect[Request] { case _ =>
```
Comment (amitksingh1490) [inline example/src/main/scala/example/PlainTextBenchmarkServer.scala]: not in the scope of this PR

File: zio-http/src/main/scala/zhttp/http/URL.scala [hunk 2]
```diff
@@ -1,167 +1,143 @@
...
-  fragment: Option[Fragment] = None,
-) { self =>
+sealed trait URL { self =>
 
-  def encode: String = URL.encode(self)
+  def getHost: Option[String]   = self.toAbsolute.host
+  def getPort: Option[Int]      = self.toAbsolute.port
+  def getScheme: Option[Scheme] = self.toAbsolute.scheme
+  def getRelative: Relative     = self.toAbsolute.relative
+  def getPath: Path             = getRelative.path
```
Comment (amitksingh1490) [inline zio-http/src/main/scala/zhttp/http/URL.scala]: drop get from the function names

File: zio-http/src/main/scala/zhttp/http/Scheme.scala [hunk 3]
```diff
@@ -48,6 +48,13 @@ object Scheme       {
     case _       => None
   }
 
+  def decode2(scheme: String): Scheme = scheme.length match {
```
Comment (amitksingh1490) [inline zio-http/src/main/scala/zhttp/http/Scheme.scala]: this implementation can introduce bugs

**Review by amitksingh1490 — CHANGES_REQUESTED** [review]

**Comment by ShrutiVerma97** [comment]
Final benchmarks:
```
[info] Benchmark                                  Mode  Cnt       Score        Error  Units
[info] URLParserBenchmark.benchmarkURLOldParser  thrpt    3  374112.291 ? 298723.960  ops/s
[info] URLParserBenchmark.benchmarkURLParser     thrpt    3  437285.905 ? 278341.194  ops/s
```
Improvement: 16%

File: zio-http/src/main/scala/zhttp/http/Request.scala [hunk 4]
```diff
@@ -78,7 +79,7 @@ trait Request extends HeaderExtension[Request] { self =>
   /**
    * Overwrites the path in the request
    */
-  def setPath(path: Path): Request = self.copy(url = self.url.copy(path = path))
+  def setPath(path: Path): Request = self.copy(url = self.url.toAbsolute.copy(urlRelative = Relative(path)))
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/Request.scala]:
```suggestion
  def setPath(path: Path): Request = self.copy(url = self.url.setPath(path))
```

File: zio-http/src/main/scala/zhttp/http/Scheme.scala (hunk 3 above)
Comment (tusharmath) [reply zio-http/src/main/scala/zhttp/http/Scheme.scala]: - Should be called as unsafeDecode
- Should be package private

File: zio-http/src/main/scala/zhttp/http/Scheme.scala [hunk 5]
```diff
@@ -48,6 +48,13 @@ object Scheme       {
...
+    case 5 => HTTPS
+    case 3 => WSS
+    case 2 => WS
+    case _ => HTTP
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/Scheme.scala]:
```suggestion
    case _ => null
```

**Pull request closed without merging**